tmt/libraries/__init__.py

```diff
@@ -0,0 +1,186 @@
+
+# A type for Beakerlib dependencies
+LibraryDependenciesType = Tuple[
+    List[tmt.base.Require], List[tmt.base.Require], List['LibraryType']
+    ]
+
+
+class LibraryError(Exception):
+    """ Used when library cannot be parsed from the identifier """
+
+
+class Library:
+    """
+    General library class
+
+    Used as parent for specific libraries like beakerlib and file
+    """
+
+    def __init__(
+            self,
+            *,
+            parent: Optional[tmt.utils.Common] = None,
+            logger: tmt.log.Logger) -> None:
+        """ Process the library identifier and fetch the library """
+        # Use an empty common class if parent not provided (for logging, cache)
+        self.parent = parent or tmt.utils.Common(logger=logger, workdir=True)
+
+        self.identifier: LibraryIdentifierType
+
+    @property
+    def hostname(self) -> str:
+        """ Get hostname from url or default to local """
+        if hasattr(self, 'url') and self.url:
+            matched = re.match(r'(?:git|http|https)://(.*?)/', self.url)
+            if matched:
+                return matched.group(1)
+        return 'local'
+
+    @property
+    def fmf_node_path(self) -> Path:
+        """ Path to fmf node """
+        assert hasattr(self, 'name')  # type check
+        if hasattr(self, 'path') and self.path:
+            return Path(self.path / self.name.strip('/'))
+        return Path(self.name)
+
+    def __str__(self) -> str:
+        """ Use repo/name for string representation """
+        assert hasattr(self, 'repo')
+        assert hasattr(self, 'name')  # type check
+        return f"{self.repo}{self.name}"
+
+
+def library_factory(
+        *,
+        identifier: LibraryIdentifierType,
+        parent: Optional[tmt.utils.Common] = None,
+        logger: tmt.log.Logger,
+        source_location: Optional[Path] = None,
+        target_location: Optional[Path] = None) -> LibraryType:
+    """ Factory function to get correct library instance """
+    if (isinstance(identifier, tmt.base.RequireSimple) or
+            isinstance(identifier, tmt.base.RequireFmfId)):
+        from .beakerlib import BeakerLib
+        library: LibraryType = BeakerLib(identifier=identifier, parent=parent, logger=logger)
+
+    # File import
+    elif isinstance(identifier, tmt.base.RequireFile):
+        assert source_location is not None
+        assert target_location is not None  # narrow type
+        from .file import File
+        library = File(
+            identifier=identifier, parent=parent, logger=logger,
+            source_location=source_location, target_location=target_location)
+
+    # Something weird
+    else:
+        raise LibraryError
+
+    # Fetch the library
+    try:
+        library.fetch()
+    except fmf.utils.RootError:
+        assert hasattr(library, 'url')  # type check
+        raise tmt.utils.SpecificationError(
+            f"Repository '{library.url}' does not contain fmf metadata.")
```

Comment:
1. please, preserve the chain of exceptions with `from ...`:

```python
    except fmf.utils.RootError as exc:
        raise tmt.utils.SpecificationError(
            f"Repository '{library.url}' does not contain fmf metadata.") from exc
```

2. I'm not sure whether `assert` is the right choice here: what if the exception is raised by a library class that does not have `url` attribute, but its `fetch()` still may fail? How about two tiers of exceptions?

```python
        if hasattr(library, 'url'):
            raise tmt.utils.SpecificationError(f"Repository '{library.url}' does not contain fmf metadata.")
        # Raise the original exception because we don't know how to provide a better error message
        raise exc
```